src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+import {BuffEvent, CastEvent} from 'fflogs'
+import _ from 'lodash'
+import Module, {dependency} from 'parser/core/Module'
+import GlobalCooldown from 'parser/core/modules/GlobalCooldown'
+import Suggestions, {TieredSuggestion} from 'parser/core/modules/Suggestions'
+import Timeline from 'parser/core/modules/Timeline'
+import React from 'react'
+
+class BuffWindowState {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+
+	constructor(start: number) {
+		this.start = start
+	}
+
+	get gcds(): number {
+		return this.rotation
+			.map(e => getDataBy(ACTIONS, 'id', e.ability.guid) as TODO)
+			.filter(a => a && a.onGcd)
+			.length
+	}
+
+	getActionCountByIds(actionsById: number[]): number {
+		return this.rotation
+			.filter(e => actionsById.includes(e.ability.guid))
+			.length
+	}
+}
+
+interface SeverityTiers {
+	[key: number]: number
+}
+
+interface BuffWindowExpectedGCDs {
+	expectedPerWindow: number
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowRequiredGCDs {
+	iconAction: ActionData
+	actions: ActionData[]
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowTrackedCooldown {
+	action: ActionData
+	expectedPerWindow: number,
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+export default class BuffWindowModule extends Module {
+	static handle: string = 'buffwindow'
+	static title: MessageDescriptor = t('core.buffwindow.title')`Buff Window`
+
+	// Implementing modules should pass the ACTION and STATUS object respectively that represents the buff window
+	protected buffAction?: ActionData
+	protected buffStatus?: StatusData
+
+	// From a practical perspective, expectedGCDs should be considered required for implementing modules.
+	// Other properties can be added as desired.  trackedBadCooldowns will show suggestions if used too many times, but will not show in the rotation tables
+	protected expectedGCDs?: BuffWindowExpectedGCDs
+	protected requiredGCDs?: BuffWindowRequiredGCDs
+	protected trackedCooldowns?: BuffWindowTrackedCooldown[]
+	protected trackedBadCooldowns?: BuffWindowTrackedCooldown[]
+
+	@dependency private suggestions!: Suggestions
+	@dependency private timeline!: Timeline
+	@dependency private globalCooldown!: GlobalCooldown
+
+	private buffWindows: BuffWindowState[] = []
+
+	private get activeBuffWindow(): BuffWindowState | undefined {
+		const lastBuffWindow = _.last(this.buffWindows)
+		if ( lastBuffWindow && lastBuffWindow.end == null ) {
+			return lastBuffWindow
+		}
+		return undefined
+	}
+
+	protected init() {
+		this.addHook('cast', {by: 'player'}, this.onCast)
+		this.addHook('applybuff', {by: 'player'}, this.onApplyBuff)
+		this.addHook('removebuff', {by: 'player'}, this.onRemoveBuff)
+		this.addHook('complete', this.onComplete)
+	}
+
+	private onCast(event: CastEvent) {
+		const action: ActionData | undefined = getDataBy(ACTIONS, 'id', event.ability.guid)
```

Comment:
if `ACTIONS` is typed as mentioned previously, _hopefully_ this won't need an explicit type on it

Reply:
I guess I don't have enough TS-foo to know how to make sure ACTIONS is typed in a way that getDataBy is returning the correct type.  Without it, I'm getting a TS error at line 102: "Property 'autoAttack' does not exist on type {insert all the 609 different types that "exist" for current ACTIONS objects}".  

Aka, I seem to be choking on the fact that not every ACTION has the autoAttack property defined, and just adding a .hasOwnProperty check doesn't seem to solve the issue.

Reply:
🤢 i had a poke around with the typings and while i _get_ what's going on, it's going to take more time to fix than I have time for immediately, primarily for the statuses.

feel free to move the types back to where they were originally or just delete them and TODO these.

Reply:
Is it going to cause a problem to have ActionData / StatusData in the .d.ts files I have currently in the data directories?  Or can we leave them there (since I do agree they make more sense there rather than glommed onto getDataBy) and if future-you fixes the typings, then this module can be updated accordingly?

Reply:
`.d.ts` for something that doesn't actually have an equivalent js file is pretty jank, but we can get away with it for now. It'll be easy enough to clean up when they get typed properly.